### PR TITLE
Fix ActiveRecord adapter name for Rails 7

### LINKED ISSUE #2924: Rails version comparison bug in Datadog::Tracing::Contrib::ActiveRecord::Utils.db_config
**Current behaviour**
`Datadog::Tracing::Contrib::ActiveRecord::Utils.adapter_name` always returns `defaultdb` for Rails version 7.0

**Expected behaviour**
`Datadog::Tracing::Contrib::ActiveRecord::Utils.adapter_name` returns the correct adapter name for Rails version 7.0

**Steps to reproduce**

[Datadog::Tracing::Contrib::ActiveRecord::Utils.db_config](https://github.com/DataDog/dd-trace-rb/blob/bdd22a40a837ad3bc9d034726753afea7f75c15a/lib/datadog/tracing/contrib/active_record/utils.rb#L115-L121) returns the configuration hash for the database.

`::Rails::VERSION::MAJOR >= 6 && ::Rails::VERSION::MINOR >= 1` is really meant to check for Rails version >= 6.1 but because of the `::Rails::VERSION::MINOR >= 1`  conditional, it doesn't pass for Rails version 7.0 (where `::Rails::VERSION::MAJOR` will be > 6, but `::Rails::VERSION::MINOR` will be < 1).

From here it tries to call `connection_pool.spec.config` (which doesn't exist for rails versions >6.1) and throws a `NoMethodError: undefined method spec for #<ActiveRecord::ConnectionAdapters::ConnectionPool...` exception.

That exception is then [caught by this line](https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/tracing/contrib/active_record/utils.rb#L110-L112), which returns an `EMPTY_CONFIG`. When it comes time to normalize the db vendor, it's nil and so defaultdb is returned here https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/tracing/contrib/utils/database.rb#L16

**Environment**

* **ddtrace version:** N/A
* **Configuration block:** N/A
* **Ruby version:** Any supported by Rails 7
* **Operating system:** Any
* **Relevant library versions:** Rails 7.0

### PR DESCRIPTION
**Pull request by TonyCTHsu**
Closes #2924 

**What does this PR do?**

Fix `Utils.adapter_name` for version > 7

1. Replace the condition with `respond_to? :db_config`, instead of Rails version check
2. Mock `Utils.adapter_name` for settings spec to avoid flaky spec
3. ~~Remove `::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver` for connection resolver~~
4. Add test with activerecord, 3.x, 4.x, 5.x, 6.0.x, 6.1.x and 7.x

### CODE REVIEW

**Review by anmarchenko — COMMENTED** [review]
looks good to me, will let someone from apm-ruby team to approve as I still lack context

File: lib/datadog/tracing/contrib/active_record/configuration/resolver.rb [hunk 1]
```diff
@@ -93,10 +93,6 @@ def parse_matcher(matcher)
             def connection_resolver
               @resolver ||= if defined?(::ActiveRecord::Base.configurations.resolve)
                               ::ActiveRecord::DatabaseConfigurations.new(active_record_configuration)
-                            elsif defined?(::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver)
```
[the remark this answers is not in the document]
Comment (TonyCTHsu) [reply lib/datadog/tracing/contrib/active_record/configuration/resolver.rb:96]: Good call out! 

I have setup test matrix in our CI to explore it a bit more. 

According to my research, `::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver` was introduced in AR 4.x , and our vendored resolver is taken from AR 3.x . 

The test matrix look like:
✅ AR 7.x (`ActiveRecord::Base.configurations.resolve`)
✅ AR 6.1.x (`ActiveRecord::Base.configurations.resolve`)
🔴  AR 6.0.x (`::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver`)
✅ AR 5.x (`::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver`)
✅ AR 4.x (`::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver`)
✅ AR 3.x ( `vendored resolver`)

This suggests  `::ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver` is working for AR 4.x and 5.x, but not 6.0.x

After removal:
✅ AR 7.x (`ActiveRecord::Base.configurations.resolve`)
✅ AR 6.1.x (`ActiveRecord::Base.configurations.resolve`)
✅ AR 6.0.x (`vendored resolver`)
✅ AR 5.x (`vendored resolver`)
✅ AR 4.x (`vendored resolver`)
✅ AR 3.x ( `vendored resolver`)

It turned out `vendored resolver` work just fine for all version prior to AR 6.1.x.

[the remark this answers is not in the document]
Comment (TonyCTHsu) [reply lib/datadog/tracing/contrib/active_record/configuration/resolver.rb:96]: It does not feel right to depend on an old vendored resolver, so I figured out it was our tests were not setting things up properly. For `6.0.x`, the resolver depends on `::ActiveRecord::DatabaseConfigurations`, which was introduced 6+.

File: Appraisals [hunk 2]
```diff
@@ -587,6 +587,15 @@ elsif ruby_version?('2.3')
   appraise 'core-old' do
     gem 'dogstatsd-ruby', '~> 4'
   end
+
+  # Somehow, I failed to install this appraisal group with Ruby 2.4
```
Comment (ekump) [inline Appraisals:591]: Should this comment be here?

**Review by ekump — APPROVED** [review]
Not that familiar with this adapter, but LGTM

File: spec/datadog/tracing/contrib/active_record/configuration/resolver_spec.rb [hunk 3]
```diff
@@ -4,9 +4,18 @@
...
+  # rubocop: disable RSpec/MultipleSubjects
+  # `::ActiveRecord.version` does not exist before 4.x
+  if ::ActiveRecord.respond_to?(:version) && ::ActiveRecord.version >= Gem::Version.new('6')
+    # ::ActiveRecord::DatabaseConfigurations` was introduced from 6+
+    require 'active_record/database_configurations'
+    subject(:resolver) { described_class.new(::ActiveRecord::DatabaseConfigurations.new(configuration)) }
+  else
+    subject(:resolver) { described_class.new(configuration) }
+  end
+  # rubocop: enable RSpec/MultipleSubjects
```
Comment (marcotc) [inline spec/datadog/tracing/contrib/active_record/configuration/resolver_spec.rb:16]:
```suggestion
  subject(:resolver) do
    if ::ActiveRecord.respond_to?(:version) && ::ActiveRecord.version >= Gem::Version.new('6')
      # ::ActiveRecord::DatabaseConfigurations` was introduced from 6+
      require 'active_record/database_configurations'
      described_class.new(::ActiveRecord::DatabaseConfigurations.new(configuration))
    else
      described_class.new(configuration)
    end
  end
```

**Review by marcotc — APPROVED** [review]

File: Appraisals (hunk 2 above)
Comment (TonyCTHsu) [reply Appraisals:591]: I supposed we are going to deprecate old ruby and Rails 3.x in the future, so did not invest too much solving this, so I left a comment instead.

**Review by anmarchenko — APPROVED** [review]

**Pull request merged**